File: src/platform/process-windows.ts
import { exec } from "node:child_process";
import { promisify } from "node:util";
import type { ProcessDiscoverer, ProcessInfo } from "./interface.js";
import { containsCaseInsensitive, parsePid } from "./process.js";

const execAsync = promisify(exec);

export class WindowsProcessDiscoverer implements ProcessDiscoverer {
  async findProcesses(): Promise<ProcessInfo[]> {
    try {
      const { stdout } = await execAsync("tasklist /FO CSV /NH");
      return this.parseTasklist(stdout);
    } catch {
      return [];
    }
  }

  async findProcessByPort(port: number): Promise<ProcessInfo | null> {
    try {
      const { stdout } = await execAsync("netstat -ano");
      const lines = stdout.split("\n");

      for (const line of lines) {
        if (line.includes(`:${port}`)) {
          const fields = line.trim().split(/\s+/);
          const pidField = fields[fields.length - 1];
          if (pidField) {
            const pid = parsePid(pidField);

            if (pid !== null && pid > 0) {
              return this.findProcessByPid(pid);
            }
          }
        }
      }

      return null;
    } catch {
      return null;
    }
  }

  async findProcessByName(name: string): Promise<ProcessInfo[]> {
    const processes = await this.findProcesses();
    return processes.filter((p) => containsCaseInsensitive(p.name, name));
  }

  async isPortInUse(port: number): Promise<boolean> {
    try {
      const { stdout } = await execAsync("netstat -ano");
      return stdout.includes(`:${port}`);
    } catch {
      return false;
    }
  }

  private async findProcessByPid(pid: number): Promise<ProcessInfo | null> {
    try {
      const { stdout } = await execAsync(
        `tasklist /FI "PID eq ${pid}" /FO CSV /NH`,
      );
      const processes = this.parseTasklist(stdout);
      return processes[0] ?? null;
    } catch {
      return null;
    }
  }

  private parseTasklist(output: string): ProcessInfo[] {
    const processes: ProcessInfo[] = [];
    const lines = output.split("\n");

    for (const line of lines) {
      if (!line.trim()) continue;

      // Format: "ImageName","PID","SessionName","Session#","MemUsage"
      const match = line.match(/"([^"]+)","(\d+)"/);
      if (match) {
        processes.push({
          name: match[1]!,
          pid: Number.parseInt(match[2]!, 10),
        });
      }
    }

    return processes;
  }
}

File: internal/platform/process_windows.go
//go:build windows

package platform

import (
	"fmt"
	"strconv"
	"strings"
)

// windowsProcessDiscoverer implements ProcessDiscoverer for Windows
type windowsProcessDiscoverer struct{}

// FindProcesses returns all running processes on Windows
func (w *windowsProcessDiscoverer) FindProcesses() ([]ProcessInfo, error) {
	output, err := runCommand("tasklist", "/FO", "CSV", "/NH")
	if err != nil {
		return nil, fmt.Errorf("failed to list processes: %w", err)
	}

	var processes []ProcessInfo
	lines := strings.Split(output, "\n")

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		// Parse CSV format: "Image Name","PID","Session Name","Session#","Mem Usage"
		fields := strings.Split(line, "\",\"")
		if len(fields) < 2 {
			continue
		}

		name := strings.Trim(fields[0], "\"")
		pidStr := strings.Trim(fields[1], "\"")
		pid := parsePID(pidStr)

		if pid > 0 {
			processes = append(processes, ProcessInfo{
				PID:  pid,
				Name: name,
				Args: "",
			})
		}
	}

	return processes, nil
}

// FindProcessByPort finds the process listening on a specific port on Windows
func (w *windowsProcessDiscoverer) FindProcessByPort(port int) (*ProcessInfo, error) {
	// First check if port is in use
	if !checkPort(port) {
		return nil, fmt.Errorf("port %d is not in use", port)
	}

	// Use netstat to find the process listening on the port
	output, err := runCommand("netstat", "-ano")
	if err != nil {
		return nil, fmt.Errorf("failed to run netstat: %w", err)
	}

	lines := strings.Split(output, "\n")
	portStr := ":" + strconv.Itoa(port)

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if !strings.Contains(line, portStr) {
			continue
		}

		// Look for LISTENING state
		if !strings.Contains(line, "LISTENING") {
			continue
		}

		// Parse netstat output: TCP    0.0.0.0:5005    0.0.0.0:0    LISTENING    12345
		fields := strings.Fields(line)
		if len(fields) < 5 {
			continue
		}

		pid := parsePID(fields[4])
		if pid == 0 {
			continue
		}

		// Get process name using tasklist
		procOutput, err := runCommand("tasklist", "/FI", "PID eq "+strconv.Itoa(pid), "/FO", "CSV", "/NH")
		if err != nil {
			continue
		}

		procLines := strings.Split(procOutput, "\n")
		for _, procLine := range procLines {
			procLine = strings.TrimSpace(procLine)
			if procLine == "" {
				continue
			}

			procFields := strings.Split(procLine, "\",\"")
			if len(procFields) < 2 {
				continue
			}

			name := strings.Trim(procFields[0], "\"")
			return &ProcessInfo{
				PID:  pid,
				Name: name,
				Args: "",
			}, nil
		}
	}

	return nil, fmt.Errorf("no process found listening on port %d", port)
}

// FindProcessByName finds processes by name on Windows
func (w *windowsProcessDiscoverer) FindProcessByName(name string) ([]ProcessInfo, error) {
	output, err := runCommand("tasklist", "/FI", "IMAGENAME eq "+name, "/FO", "CSV", "/NH")
	if err != nil {
		return nil, fmt.Errorf("failed to find process by name: %w", err)
	}

	var processes []ProcessInfo
	lines := strings.Split(output, "\n")

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		// Parse CSV format: "Image Name","PID","Session Name","Session#","Mem Usage"
		fields := strings.Split(line, "\",\"")
		if len(fields) < 2 {
			continue
		}

		processName := strings.Trim(fields[0], "\"")
		pidStr := strings.Trim(fields[1], "\"")
		pid := parsePID(pidStr)

		if pid > 0 {
			processes = append(processes, ProcessInfo{
				PID:  pid,
				Name: processName,
				Args: "",
			})
		}
	}

	if len(processes) == 0 {
		return nil, fmt.Errorf("no process found with name %s", name)
	}

	return processes, nil
}
